Simd

FIVIL opened the pull request:
This PR contains a simple tutorial and examples for SIMD-enabled types in .NET.
 fixes https://github.com/dotnet/docs/issues/12660

/cc @rpetrusha

FIVIL committed:
mera added

FIVIL committed:
basic

FIVIL committed:
vector and matrix

FIVIL committed:
remarks

FIVIL committed:
fixing headers problem

mairaw commented:
Also tagging @tannergooding and @ViktorHofer for a technical review.

tannergooding commented on docs/standard/simd.md:
```diff
@@ -0,0 +1,104 @@
+
+# Overview
+
+SIMD (Single instruction, multiple data) provides the functionality to do multiple processes on a single core to achieve peak performance. In .NET there is set of SIMD-enabled types under <xref:System.Numerics> namespace. SIMD operations can be parallelized at the hardware level. That increases the throughput of the vectorized computations, which are common in mathematical, scientific, and graphi ...
```
This first sentence could be reworded.

Maybe something like:
> SIMD (Single instruction, multiple data) provides hardware support for performing an operation on multiple pieces of data, in parallel, using a single instruction.

tannergooding replied:
```suggestion
SIMD (Single instruction, multiple data) provides hardware support for performing an operation on multiple pieces of data, in parallel, using a single instruction. In .NET there is set of SIMD-enabled types under <xref:System.Numerics> namespace. SIMD operations can be parallelized at the hardware level. That increases the throughput of the vectorized computations, which are common in mathematical, scientific, and graphics apps.
```

tannergooding commented on docs/standard/simd.md:
```diff
@@ -0,0 +1,104 @@
+
+# Overview
+
+SIMD (Single instruction, multiple data) provides the functionality to do multiple processes on a single core to achieve peak performance. In .NET there is set of SIMD-enabled types under <xref:System.Numerics> namespace. SIMD operations can be parallelized at the hardware level. That increases the throughput of the vectorized computations, which are common in mathematical, scientific, and graphi ...
```
It is probably worthwhile to callout both the `System.Numerics` types and the System.Runtime.Intrinsics namespaces. 
The former provides SIMD accelerated functions (in `Vector2/3/4` and other types) and more direct SIMD access (`Vector<T>`) which may have a software fallback on some platforms (such as `arm32`).

The latter provides direct hardware access and no software fallback; so support must be explicitly queried and fallbacks need to be provided by the developer.

FIVIL replied:
Actually, I have just started to work with new `System.Runtime.Intrinsic` in .net core 3 and I will complete these documents based on this [blog post](https://devblogs.microsoft.com/dotnet/hardware-intrinsics-in-net-core/) of yours, very soon

tannergooding commented on docs/standard/simd.md:
```diff
@@ -0,0 +1,104 @@
+
+## .NET SIMD-enabled types
+
+The .NET SIMD-enabled types include the following:
```
I'm not sure saying they are `SIMD-enabled` is correct. Its that certain exposed APIs on the types may be SIMD accelerated.

tannergooding replied:
```suggestion
## .NET SIMD accelerated types

The .NET SIMD accelerated types include the following:
```

tannergooding commented on docs/standard/simd.md:
```diff
@@ -0,0 +1,104 @@
+
+## How to use SIMD?
+
+ Before executing custom SIMD algorithms, it is possible to check if the host machine supports SIMD using <xref:System.Numerics.Vector.IsHardwareAccelerated?displayProperty=nameWithType>, which returns a <xref:System.Boolean>.
```
I think this should be explained that `IsHardwareAccelerated` returning `true` only indicates that some functionality may be hardware accelerated. It doesn't provide any guarantees of what functions are accelerated (you may even end up in a scenario where one function is accelerated but another function is not).

FIVIL replied:
This makes sense, I will make some changes.
@tannergooding  Tnx for your suggestions, I will make some changes based on your review

tdykstra commented:
@FIVIL Do you still plan to address the feedback on this PR?

FIVIL commented:
@tdykstra  Sure, I am currently testing new `Intrinsics`  API in .Net Core 3+. And as soon as I grasp a deep understanding of all the new APIS I will complete this tutorial and address the issues mentioned in the feedback

Thraka committed:
Update simd.md

- Updated the date
- Changed terms from **SIMD-enabled** to **SIMD-accelerated**
- Fixed any spelling errors
- Fixed code spacing
- Adjusted markdown structure
- Brought acro score from 77 to 92

Thraka commented:
@tannergooding @FIVIL 

Please re-review. I've updated the submission with the following changes:

- Updated the date
- Changed terms from **SIMD-enabled** to **SIMD-accelerated**
- Fixed any spelling errors
- Fixed code spacing
- Adjusted markdown structure
- Brought acro score from 77 to 92 with minor grammar changes

tannergooding commented on docs/standard/simd.md:
```diff
@@ -0,0 +1,103 @@
+
+- The <xref:System.Numerics.Vector2>, <xref:System.Numerics.Vector3>, and <xref:System.Numerics.Vector4> types, which represent vectors with 2, 3, and 4 <xref:System.Single> values.
+
+- Two matrix types, <xref:System.Numerics.Matrix3x2>, which represents a 3x2 matrix, and <xref:System.Numerics.Matrix4x4>, which represents a 4x4 matrix.
```
Should we call out that this is xref:System.Single values, like we did for Vector2/3/4?

Likewise for Plane/Quaternion?

Thraka replied:
Would you be willing to write that content? It's a bit out of my area of expertise.

tannergooding replied:
```suggestion
- Two matrix types, <xref:System.Numerics.Matrix3x2>, which represents a 3x2 matrix, and <xref:System.Numerics.Matrix4x4>, which represents a 4x4 matrix of <xref:System.Single> values.
```

tannergooding commented on docs/standard/simd.md:
```diff
@@ -0,0 +1,103 @@
+
+- The <xref:System.Numerics.Quaternion> type, which represents a vector that is used to encode three-dimensional physical rotations.
+
+- The <xref:System.Numerics.Vector%601> type, which represents a vector of a specified numeric type and provides a broad set of operators that benefit from SIMD support. The count of a <xref:System.Numerics.Vector%601> instance is fixed, but its value <xref:System.Numerics.Vector%601.Count%2A?displayProperty=nameWithType> depends on the CPU of the machine running the code.
```
This could probably be reworded to better indicate it is fixed for the lifetime of an application but the count is dependent on the underlying hardware.

tannergooding replied:
```suggestion
- The <xref:System.Numerics.Vector%601> type, which represents a vector of a specified numeric type and provides a broad set of operators that benefit from SIMD support. The count of a <xref:System.Numerics.Vector%601> instance is fixed for the lifetime of an application, but its value <xref:System.Numerics.Vector%601.Count%2A?displayProperty=nameWithType> depends on the CPU of the machine running the code.
```

tannergooding commented on docs/standard/simd.md:
```diff
@@ -0,0 +1,103 @@
+
+## How to use SIMD?
+
+Before executing custom SIMD algorithms, it's possible to check if the host machine supports SIMD by using <xref:System.Numerics.Vector.IsHardwareAccelerated?displayProperty=nameWithType>, which returns a <xref:System.Boolean>. This doesn't guarantee that SIMD-acceleration is enabled for a specific type, but is an indicator that it's supported by some types.
```
Maybe add in that it doesn't guarantee for a specific type or even for a particular method of a supported type?

That is, `Vector<float>` could be partially accelerated and so `Vector<float>.Method1` might be accelerated while `Vector<float>.Method2` might not be.
It is all dependent on the underlying hardware.

tannergooding commented on docs/standard/simd.md:
```diff
@@ -0,0 +1,103 @@
+
+## Remarks
+
+SIMD is more likely to remove one bottleneck and expose the next, for example memory throughput. In general the performance benefit of using SIMD varies depending on the specific scenario, and in some cases it can even perform worse than simpler non-SIMD equivalent code.
```
End with a note that users should benchmark and profile their code to identify hot spots and ensure that changes are meeting their performance goals/needs?

Can we link to https://github.com/dotnet/BenchmarkDotNet?

tannergooding commented:
Overall this looks really good, thanks for investing the time into this. I just had a couple minor questions/feedback items 😄

Thraka commented:
@tannergooding

I tried working on rewording it but I feel like I'll misrepresent the details. Can you submit a suggestion where you want changes? This way you can reword them easily without having to fork and submit a huge update

![image](https://user-images.githubusercontent.com/2672110/79255207-c108be00-7e3a-11ea-8a99-f094848c643d.png)

Thraka commented:
@tannergooding Ping on my last comment :)

tannergooding commented on docs/standard/simd.md:
```diff
@@ -0,0 +1,104 @@
+
+## How to use SIMD?
+
+ Before executing custom SIMD algorithms, it is possible to check if the host machine supports SIMD using <xref:System.Numerics.Vector.IsHardwareAccelerated?displayProperty=nameWithType>, which returns a <xref:System.Boolean>.
```
```suggestion
 Before executing custom SIMD algorithms, it is possible to check if the host machine supports SIMD using <xref:System.Numerics.Vector.IsHardwareAccelerated?displayProperty=nameWithType>, which returns a <xref:System.Boolean>. However, this does not indicate all members are accelerated just that some members may be. Which members are accelerated are hardware dependent.
```

tannergooding commented on docs/standard/simd.md:
```diff
@@ -0,0 +1,104 @@
+
+- Two matrix types, <xref:System.Numerics.Matrix3x2>, which represents a 3x2 matrix, and <xref:System.Numerics.Matrix4x4>, which represents a 4x4 matrix.
+
+- The <xref:System.Numerics.Plane> type, which represents a plane in three-dimensional space.
```
```suggestion
- The <xref:System.Numerics.Plane> type, which represents a plane in three-dimensional space using <xref:System.Single> values.
```

tannergooding commented on docs/standard/simd.md:
```diff
@@ -0,0 +1,104 @@
+
+- The <xref:System.Numerics.Plane> type, which represents a plane in three-dimensional space.
+
+- The <xref:System.Numerics.Quaternion> type, which represents a vector that is used to encode three-dimensional physical rotations.
```
```suggestion
- The <xref:System.Numerics.Quaternion> type, which represents a vector that is used to encode three-dimensional physical rotations using <xref:System.Single> values.
```

Thraka committed:
Update docs/standard/simd.md

Co-Authored-By: Tanner Gooding <tagoo@outlook.com>

Thraka committed:
Update docs/standard/simd.md

Co-Authored-By: Tanner Gooding <tagoo@outlook.com>

Thraka committed:
Apply suggestions from code review

Co-Authored-By: Tanner Gooding <tagoo@outlook.com>

Thraka reviewed (approved):
Something is going on with github. I can't accept suggestions and I see old versions when trying to review. I did a lot of rewrite work about a month ago and maybe that is causing problems? I think it's good to go though.
